src/WRings.sol

```diff
@@ -0,0 +1,124 @@
+    }
+
+    /// @notice User specifies amount of asset token to deposit in return for vault shares
+    /// @param assets Number of asset tokens to deposit
+    /// @param receiver Address that vault tokens will be minted to
+    /// @return shares Number of vault shares to return to the user
+    function deposit(uint256 assets, address receiver) external  returns(uint256) {
+        uint256 shares = WMul.wmul(assets * 10**9, exchangeRate) / 10**9;
+        mint(receiver, shares);
+        iRings.transferFrom(msg.sender, address(this), assets);
+        emit Deposit(msg.sender, receiver, assets, shares);
+        return shares;
+    }
+
+    /// @notice User specifies the desired amount of vault shares in exchange for asset tokens
+    /// @param shares Number of vault tokens desired by the user
+    /// @param receiver Address that vault Tokens will be minted to
+    /// @return assets Number of asset tokens required to mint desired amount of shares
+    function mint(uint256 shares, address receiver) external  returns(uint256) {
+        uint256 assets = WDiv.wdiv(shares * 10**9, exchangeRate) / 10**9;
+        mint(receiver, shares);
+        iRings.transferFrom(msg.sender, address(this), assets);
+        emit Deposit(msg.sender, receiver, assets, shares);
+        return assets;
+    }
+
+    /// @notice User specifies amount of asset tokens to withdraw from the vault
+    /// @dev
+    /// @param assets Number of asset token that user desires to withdraw from the vault
+    /// @param receiver Address that will receive the withdrawn tokens
+    /// @param owner Address that currently owns the vault tokens
+    /// @return shares Number of shares required to be exchange for desired amount of asset tokens
+    function withdraw(uint256 assets, address receiver, address owner) external  returns(uint256){
+        uint256 shares = WMul.wmul(assets * 10**9, exchangeRate) / 10**9;
+        require(_balanceOf[owner] >= shares, "Owner holds less than required amount of vault shares for requested assets");
+        burn(owner, shares);
+        iRings.transfer(receiver, assets);
+        emit Withdraw(msg.sender, receiver, owner, assets, shares);
+        return shares;
+    }
+
+    /// @notice User specifies amount of share tokens to exchange for asset tokens held in the vault
+    /// @dev
+    /// @param shares Number of vault tokens to be exchanged
+    /// @param receiver Address that will receive withdrawn tokens
+    /// @param owner Address that currently owns the vault tokens
+    /// @return assets Number of asset tokens received in exchange for the specified vault tokens
+    function redeem(uint256 shares, address receiver, address owner) external returns(uint256) {
+        require(_balanceOf[owner] >= shares, "Owner holds less than specified amount of vault shares");
+        uint256 assets = WDiv.wdiv(shares * 10**9, exchangeRate) / 10**9;
+        burn(owner, shares);
+        iRings.transfer(receiver, assets);
```

Comment:
Now that you know how to attach library methods to types, look out for `TransferHelper.sol` in the yiled-utils-v2 repo and use it to make sure all your transfers revert even if the underlying ERC20 transfer returns `false` instead of reverting.